pkg/operator/reports_test.go

```diff
@@ -92,3 +101,186 @@ func TestGetNextReportPeriod(t *testing.T) {
+				},
+			}, schedule, false),
+			expectRunning: false,
+		},
+		{
+			name: "spec.Status set to v1.ConditionTrue returns false",
+			report: testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportStart, reportEnd, v1alpha1.ReportStatus{
+				Conditions: []metering.ReportCondition{
+					*util.NewReportCondition(metering.ReportRunning, v1.ConditionTrue, util.ReportFinishedReason, "mock-msg"),
+				},
+			}, schedule, false),
+			expectRunning: false,
+		},
+	}
+
+	for _, testCase := range testTable {
+		var mockLogger = logrus.New()
+		testCase := testCase
+
+		t.Run(testCase.name, func(t *testing.T) {
+			runningCond := isReportFinished(mockLogger, testCase.report)
+			assert.Equalf(t, runningCond, testCase.expectRunning, "expected the report '%s' would return '%t', but got '%t'", testCase.name, testCase.expectRunning, runningCond)
+		})
+	}
+}
+
+func TestValidateReport(t *testing.T) {
+	const (
+		testNamespace           = "default"
+		testReportName          = "test-report"
+		testQueryName           = "test-query"
+		testGenerationQueryName = "test-GQN"
+	)
+
+	reportStart := &time.Time{}
+	reportEndTmp := reportStart.AddDate(0, 1, 0)
+	reportEnd := &reportEndTmp
+
+	testTable := []struct {
+		name         string
+		report       *metering.Report
+		expectErr    bool
+		expectErrMsg string
+	}{
+		{
+			name:         "empty spec.generationQuery returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, "", reportStart, reportEnd, v1alpha1.ReportStatus{}, nil, false),
+			expectErr:    true,
+			expectErrMsg: "must set spec.generationQuery",
+		},
+		{
+			name:         "spec.ReportingStart > spec.ReportingEnd returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportEnd, reportStart, v1alpha1.ReportStatus{}, nil, false),
+			expectErr:    true,
+			expectErrMsg: fmt.Sprintf("spec.reportingEnd (%s) must be after spec.reportingStart (%s)", reportStart.String(), reportEnd.String()),
+		},
+		{
+			name:         "spec.ReportingEnd is unset and spec.RunImmediately is set returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportStart, nil, v1alpha1.ReportStatus{}, nil, true),
+			expectErr:    true,
+			expectErrMsg: "spec.reportingEnd must be set if report.spec.runImmediately is true",
+		},
+		{
+			name:         "spec.GenerationQueryName does not exist returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportStart, reportEnd, v1alpha1.ReportStatus{}, nil, false),
+			expectErr:    true,
+			expectErrMsg: fmt.Sprintf("ReportGenerationQuery %s does not exist", testGenerationQueryName),
+		},
+	}
+
+	// hardcore the report spec's generation query name for tests that need that
+	testTable[3].report.Spec.GenerationQueryName = testGenerationQueryName
+	ds1 := testhelpers.NewReportDataSource("datasource1", testNamespace)
```

Comment:
All of the queries, datasources and the `.*Stores` setup i think should be above where we define the test cases so that it's done prior to referencing things in the cases.

Reply:
That makes sense. Updated.